### ISSUE TITLE
Segfault: Type checker fails to detect single negative numbers if record type is expected

### ISSUE THREAD

**Issue by limasierra** [souffle-lang/souffle#238]
Hi all,

Nasty little bug which can lead to segfaults. Consider the following code:

```
.type pair = [ X:number, Y:number ]

.decl MyRelation ( P:pair ) output

MyRelation([7,7]). // Works
//MyRelation(7). // Works - Error detected
MyRelation(-7). // ERROR: Negative numbers bypass checks

.decl TheDoom ( First:number ) output

TheDoom(X) :- MyRelation([X,_]). // Unpack that -7, hehe...
```

I think the comments make the problem quite clear. Execute that piece of code with `souffle -D- analysis.dl` and enjoy your segmentation fault :)

Other observations: `MyRelation(0)` and `MyRelation(nil)` are handled correctly. Numbers below 0 may avoid segfaults if they can be unpacked, but that leads to faulty results of course.

Kind regards,
Laurent

**Comment by b-scholz** [comment]
Well spotted. We will look into this.

**Comment by lyndonhenry** [comment]
Note that this bug occurs for any arithmetic expression, for example `MyRelation(1+1).` bypasses checks also; however in this instance it does not segfault. Curiously, `MyRelation([7,7,7]).` causes an error to be detected, yet `MyRecord([7])` does not. This is clearly indicative of a problem with the type system. I've included an example below to illustrate current behavior, and also added it to the testsuite. 

`
//
// Record Type
//
// This test checks that record types are handled correctly.
// It was created for issue #238 "Segfault: Type checker fails to detect single negative numbers if record type is expected".
//

.type record = [ X:number, Y:number ]

.decl MyRelation( X:record )

// No Error
MyRelation(nil).
MyRelation([1, 1]).
MyRelation([-1, -1]).

// Error
MyRelation(0).
MyRelation(1).
MyRelation("1").
MyRelation(-1). // TODO: error not detected and segfaults
MyRelation(1+1). // TODO: error not detected
MyRelation(1, 1).
MyRelation(-1, -1).
MyRelation([1]). // TODO: error not detected
MyRelation([1, 1, 1]).

.decl YourRelation( X:number ) output

YourRelation(x) :- MyRelation([x, _]).
`

**Comment by lyndonhenry** [comment]
Similarly, the following runs without detecting any errors:

```
.decl SymbolRelation( X:symbol )

SymbolRelation(1+1).
```

**Comment by lyndonhenry** [comment]
I have now solved this (#238) issue locally and will be doing a pull request soon, it turned out to be due to arithmetic expressions in general being confused for record types. However the other issue mentioned in the comments above has to do with  the type checker failing to give errors for records of lesser size than is required. As this is an unrelated problem, I have opened a new issue (#293) for it.
